qutebrowser/browser/mhtml.py

```diff
@@ -0,0 +1,534 @@
+            msg.set_type(f.content_type)
+        msg.set_payload(f.content)
+        f.transfer_encoding(msg)
+        return msg
+
+
+class _Downloader():
+
+    """A class to download whole websites.
+
+    Attributes:
+        web_view: The QWebView which contains the website that will be saved.
+        dest: Destination filename.
+        writer: The MHTMLWriter object which is used to save the page.
+        loaded_urls: A set of QUrls of finished asset downloads.
+        pending_downloads: A set of unfinished (url, DownloadItem) tuples.
+        _finished: A flag indicating if the file has already been written.
+        _used: A flag indicating if the downloader has already been used.
+        _win_id: The window this downloader belongs to.
+    """
+
+    def __init__(self, web_view, dest, win_id):
+        self.web_view = web_view
+        self.dest = dest
+        self.writer = None
+        self.loaded_urls = {web_view.url()}
+        self.pending_downloads = set()
+        self._finished = False
+        self._used = False
+        self._win_id = win_id
+
+    def run(self):
+        """Download and save the page.
+
+        The object must not be reused, you should create a new one if
+        you want to download another page.
+        """
+        if self._used:
+            raise ValueError("Downloader already used")
+        self._used = True
+        web_url = self.web_view.url()
+        web_frame = self.web_view.page().mainFrame()
+
+        self.writer = MHTMLWriter(
+            web_frame.toHtml().encode('utf-8'),
+            content_location=urlutils.encoded_url(web_url),
+            # I've found no way of getting the content type of a QWebView, but
+            # since we're using .toHtml, it's probably safe to say that the
+            # content-type is HTML
+            content_type='text/html; charset="UTF-8"',
+        )
+        # Currently only downloading <link> (stylesheets), <script>
+        # (javascript) and <img> (image) elements.
+        elements = web_frame.findAllElements('link, script, img')
+
+        for element in elements:
+            element = webelem.WebElementWrapper(element)
+            # Websites are free to set whatever rel=... attribute they want.
+            # We just care about stylesheets and icons.
+            if not _check_rel(element):
+                continue
+            if 'src' in element:
+                element_url = element['src']
+            elif 'href' in element:
+                element_url = element['href']
+            else:
+                # Might be a local <script> tag or something else
+                continue
+            absolute_url = web_url.resolved(QUrl(element_url))
+            self.fetch_url(absolute_url)
+
+        styles = web_frame.findAllElements('style')
+        for style in styles:
+            style = webelem.WebElementWrapper(style)
+            if 'type' in style and style['type'] != 'text/css':
+                continue
```

Comment:
[The Mozilla Developer Network](https://developer.mozilla.org/de/docs/Web/HTML/Element/style says

> type
>    This attribute defines the styling language as a MIME type (charset should not be specified). This attribute is optional and default to text/css if it's missing.